pkg/cli/sanitize/hide_secrets.go

```diff
@@ -0,0 +1,207 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sanitize
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"helm.sh/helm/v3/pkg/release"
+
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	hiddenSecretValue = "[HIDDEN]"
+)
+
+// HideManifestSecrets sanitizes release manifest and replaces it.
+// Manifest cannot be applied after this operation.
+// Indentation and extra spaces in Secret's `data` and `stringData` sections can be impacted.
+func HideManifestSecrets(r *release.Release) error {
+	if r == nil {
+		return nil
+	}
+	manifest, err := hideSecrets(r.Manifest)
+	if err != nil {
+		return err
+	}
+
+	r.Manifest = manifest
+
+	return nil
+}
+
+// hideSecrets replaces values in Secrets in the chart manifest with
+// `[HIDDEN]` value.
+func hideSecrets(manifest string) (string, error) {
+	resources := strings.Split(manifest, "\n---")
+	outRes := make([]string, 0, len(resources))
+
+	for _, r := range resources {
+		var resourceMap map[string]interface{}
+		err := yaml.Unmarshal([]byte(r), &resourceMap)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to unmarshal %q resource", tryToGetName(resourceMap))
+		}
+
+		if isSecret(resourceMap) {
+			rs, err := hideSecretData(r)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to hide %q Secret data", tryToGetName(resourceMap))
+			}
+			r = rs
+		}
+
+		outRes = append(outRes, r)
+	}
+
+	return strings.Join(outRes, "\n---"), nil
+}
+
+func tryToGetName(resourceMap map[string]interface{}) string {
+	metadata, ok := resourceMap["metadata"].(map[string]interface{})
+	if !ok {
+		return ""
+	}
+
+	name, ok := metadata["name"].(string)
+	if !ok {
+		return ""
+	}
+
+	return name
+}
+
+func isSecret(resource map[string]interface{}) bool {
+	kind, ok := resource["kind"].(string)
+	if !ok || kind != "Secret" {
+		return false
+	}
+
+	apiVersion, ok := resource["apiVersion"].(string)
+	if !ok || apiVersion != "v1" {
+		return false
+	}
+
+	return true
+}
+
+func hideSecretData(raw string) (string, error) {
+	lines := strings.Split(raw, "\n")
+	outLines := make([]string, 0, len(lines))
+
+	// To minimize impact of empty lines and custom indentation
+	// we only marshal `data` and `secretData` sections of the Secrets
+	for i := 0; i < len(lines); i++ {
+		line := lines[i]
+		outLines = append(outLines, line)
+
+		if strings.HasPrefix(line, "data:") || strings.HasPrefix(line, "stringData:") {
+			endLine := findSectionEnd(lines, i+1)
+			sanitizedLines, err := sanitizeDataYaml(lines[i+1 : endLine+1])
+			if err != nil {
+				return "", errors.Wrap(err, "failed to sanitize Secret data")
+			}
+			outLines = append(outLines, sanitizedLines...)
+
+			i = endLine
+		}
+	}
+
+	return strings.Join(outLines, "\n"), nil
+}
+
+func sanitizeDataYaml(yamlLines []string) ([]string, error) {
+	yamlData := strings.Join(yamlLines, "\n")
+	data := &yaml.Node{}
+	err := yaml.Unmarshal([]byte(yamlData), data)
```

Comment:
This seems... incredibly dangerous. Not because we're parsing the file, but because we're parsing it and treating it as valid data.

I would highly urge you to use existing Kubernetes client libraries used in other parts of Helm to parse and validate manifests rather than calling raw `yaml.Unmarshal` APIs. Handling and parsing raw data without prior validation/sanitization has lead to many security issues in the past.

Have a look at https://pkg.go.dev/helm.sh/helm/v3@v3.6.3/pkg/kube for inspiration.